### PR TITLE
feat(dashboard): Change release widget to include all release data [SEN-237]

### PR DESCRIPTION
**Pull request by billyvg**
* Adds an additional query field `condition_fields` to discover query endpoint that current only
 supports an "if" clause when selecting a column. i.e. we can do a query like
 "select [if <condition> is true, use <column> value, else use literal value] as alias"

 * adds support for a "<column> is in <list>" condition

* Updates release widget to use releases that the rest of the dashboard has access to
  and fetches event stats for each of those releases (and group into "Other" category otherwise)

Fixes SEN-237

### CODE REVIEW

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 1]
```diff
@@ -131,6 +136,8 @@ def validate_aggregations(self, attrs, source):
         return attrs
 
     def get_array_field(self, field):
+        if not isinstance(field, six.string_types):
```
Comment (billyvg) [inline src/sentry/api/endpoints/organization_discover_query.py]: this isn't needed anymore I think

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 2]
```diff
@@ -303,7 +310,8 @@ def post(self, request, organization):
 
         has_aggregations = len(serialized.get('aggregations')) > 0
 
-        selected_columns = [] if has_aggregations else serialized.get('fields')
+        selected_columns = serialized.get(
```
Comment (billyvg) [inline src/sentry/api/endpoints/organization_discover_query.py]: These conditions can also contain an aggregation

File: src/sentry/static/sentry/app/views/organizationDashboard/data/queries/eventsByRelease.jsx [hunk 3]
```diff
@@ -5,14 +5,71 @@ import {t} from 'app/locale';
 
 const eventsByRelease = {
   name: t('Events by Release'),
-  fields: ['release'],
+  fields: ['__release'],
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationDashboard/data/queries/eventsByRelease.jsx]: This is required for the frontend when parsing results -- we don't forward this when making discover query. Currently, we need to declare it here because this query gets passed around when generating results.

File: src/sentry/static/sentry/app/views/organizationDashboard/discoverQuery.jsx [hunk 4]
```diff
@@ -109,7 +148,7 @@ class DiscoverQuery extends React.Component {
 
   async fetchData() {
     // Reset query builder
-    this.resetQueries();
+    // this.resetQueries();
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationDashboard/discoverQuery.jsx]: Resetting queries will clear our generated query :/

File: src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx [hunk 5]
```diff
@@ -220,9 +220,10 @@ function getDataWithKeys(data, query, options = {}) {
       return row;
     }
 
-    const key = fields.length
-      ? fields.map(field => getLabel(row[field], options)).join(',')
-      : aggregate;
+    const key =
+      fields && fields.length
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: Can revert this change

File: src/sentry/utils/snuba.py [hunk 6]
```diff
@@ -301,14 +301,22 @@ def transform_aliases_and_query(**kwargs):
...
+        else:
+            name = get_snuba_column_name(col)
+            selected_columns[idx] = name
+            translated_columns[name] = col
 
     for (idx, col) in enumerate(groupby):
-        name = get_snuba_column_name(col)
-        groupby[idx] = name
-        translated_columns[name] = col
+        if not col.startswith('__'):
```
Comment (billyvg) [inline src/sentry/utils/snuba.py]: Not sure the best way to handle this -- basically we our `if` condition results has a special name that we need to refer to. `get_snuba_column_name` will potentially change to `tags[<name>]`

Comment (billyvg) [reply src/sentry/utils/snuba.py]: Any thoughts here @lynnagara?

File: src/sentry/static/sentry/app/views/organizationDashboard/data/queries/eventsByRelease.jsx [hunk 7]
```diff
@@ -5,14 +5,71 @@ import {t} from 'app/locale';
...
+// ],
+// '_release',
+// ],
+// ],
+// limit: 1000,
+// to_date: '2019-03-13T01:19:07',
+// granularity: 86400,
+// conditions: [['project_id', 'IN', [2]]],
+// groupby: ['time', 'tags[_release]'],
+// };
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/organizationDashboard/data/queries/eventsByRelease.jsx]: What are these for?

Comment (billyvg) [reply src/sentry/static/sentry/app/views/organizationDashboard/data/queries/eventsByRelease.jsx]: still a WIP

File: src/sentry/utils/snuba.py [hunk 8]
```diff
@@ -298,15 +298,66 @@ def transform_aliases_and_query(**kwargs):
     conditions = kwargs['conditions'] or []
     filter_keys = kwargs['filter_keys']
 
+    def parse_selected_column_condition(column):
```
Comment (billyvg) [inline src/sentry/utils/snuba.py]: Can you take a look at this @alex-hofsteede ?

File: src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx (hunk 5 above)
Comment (lynnagara) [reply src/sentry/static/sentry/app/views/organizationDiscover/result/utils.jsx]: Yeah, it's not a necessary check, it'll never be falsy otherwise there's big problems

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 9]
```diff
@@ -303,7 +308,8 @@ def post(self, request, organization):
...
+            'condition_fields') or [] if has_aggregations else serialized.get('fields')
```
Comment (lynnagara) [inline src/sentry/api/endpoints/organization_discover_query.py]: Can we merge condition_fields and fields if both are provided rather than ignore 1 over the other

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 10]
```diff
@@ -41,6 +41,11 @@ class DiscoverQuerySerializer(serializers.Serializer):
         required=False,
         allow_null=True,
     )
+    condition_fields = ListField(
```
Comment (lynnagara) [inline src/sentry/api/endpoints/organization_discover_query.py]: This api uses camel case - i.e. conditionFields

File: src/sentry/utils/snuba.py [hunk 11]
```diff
@@ -298,13 +298,67 @@ def transform_aliases_and_query(**kwargs):
...
+            if_condition = column[1][0]
+            new_if_condition = []
+
+            if isinstance(if_condition, list) and if_condition[0] == 'in':
+                new_if_condition.append(if_condition[0])
+                in_clause = if_condition[1]
+
+                # ["in", [<column name> , <str literal e.g. "tuple">, <list>]]
+                if in_clause[1] != 'tuple' or not isinstance(in_clause[2], list):
+                    # This is an invalid query, raise something
```
Comment (lynnagara) [inline src/sentry/utils/snuba.py]: `raise InvalidQuery`?

File: src/sentry/utils/snuba.py (hunk 8 above)
Comment (alex-hofsteede) [reply src/sentry/utils/snuba.py]: Oops, yeah just saw this, can review tomorrow.

File: src/sentry/utils/snuba.py [hunk 12]
```diff
@@ -298,13 +298,67 @@ def transform_aliases_and_query(**kwargs):
...
+                    raise
```
Comment (markstory) [inline src/sentry/utils/snuba.py]: What are you raising? You could use `sentry.search.utils.InvalidQuery`

Comment (billyvg) [reply src/sentry/utils/snuba.py]: Yeah I wasn't sure what to raise here

File: src/sentry/utils/snuba.py [hunk 13]
```diff
@@ -298,13 +298,67 @@ def transform_aliases_and_query(**kwargs):
...
 
+    def parse_selected_column_condition(column):
+        new_column = [column[0]]
+        new_condition = []
+
+        if column[0] == 'if':
+            if_condition = column[1][0]
+            new_if_condition = []
+
+            if isinstance(if_condition, list) and if_condition[0] == 'in':
```
Comment (alex-hofsteede) [inline src/sentry/utils/snuba.py]: This seems really special-purpose and fragile. I think it should be possible to use an approach similar to https://github.com/getsentry/snuba/blob/d962a354338abc29f9331c927d16ba5e560f1d59/snuba/util.py#L281-L296 to find column names inside nested functions and translate them in-place there.

Comment (billyvg) [reply src/sentry/utils/snuba.py]: Yep agreed - I couldn't think of a way to make this generic enough. One issue is that we assume if a column name is not in `SENTRY_SNUBA_MAP` then we transform it: https://github.com/getsentry/sentry/blob/3fd534be6b269f5b6fe58ac8b9e669c64f8ad729/src/sentry/utils/snuba.py#L270

Also our "col in list" expression has a different format afaict, where it's ["in", [<column name>, "tuple", <list>]] where it seems like most expressions treat the first item in a list as a keyword and the rest as arguments?

File: src/sentry/utils/snuba.py (hunk 12 above)
Comment (markstory) [reply src/sentry/utils/snuba.py]: Seems like the user has requested an InvalidQuery.

Comment (billyvg) [reply src/sentry/utils/snuba.py]: This no longer applies, we now do the bare minimum in parsing so that snuba handles the bulk of it.

**Review by alex-hofsteede — APPROVED** [review]
Yeah this looks ok if it does the needful. 

I definitely have a little bit of anxiety about the way we're leaking through some of the snuba implementation details up to this level. We should probably do a bit of a refactor at some point to try and make the snuba query API work better so that clients don't have to generate an expression, and then parse it again to translate some of the values in it, before sending to snuba. 

Is there a way that snuba could probide better `ALIAS` type functionality to prevent having to do these translations on the client side?

**Comment by billyvg** [comment]
Agree we should look into cleaning this up a bit.
